File: app/schemas/report.py
"""Reporting schemas (API_SPECIFICATION.md §14 — JSON only, 4 report types)."""
from __future__ import annotations

from decimal import Decimal

from pydantic import BaseModel, Field


class ChartPoint(BaseModel):
    label: str
    value: Decimal


# --- Sales ---------------------------------------------------------------
class PaymentStatusBucket(BaseModel):
    status: str  # paid | partial | unpaid
    count: int
    revenue: Decimal


class SalesReport(BaseModel):
    total_revenue: Decimal
    total_count: int
    by_payment_status: list[PaymentStatusBucket] = Field(default_factory=list)
    by_day: list[ChartPoint] = Field(default_factory=list)


# --- Inventory (current store_stock levels, by product) ------------------
class InventoryReportRow(BaseModel):
    product_id: int
    name: str
    sku: str
    quantity: Decimal


class InventoryReport(BaseModel):
    rows: list[InventoryReportRow] = Field(default_factory=list)
    count: int


# --- Debts (outstanding, grouped by customer and by status) --------------
class DebtByCustomer(BaseModel):
    customer_id: int
    full_name: str
    remaining: Decimal


class DebtByStatus(BaseModel):
    status: str  # active | overdue
    count: int
    remaining: Decimal


class DebtReport(BaseModel):
    by_customer: list[DebtByCustomer] = Field(default_factory=list)
    by_status: list[DebtByStatus] = Field(default_factory=list)
    total_remaining: Decimal


# --- Expenses (totals grouped by expense_type and by date) ---------------
class ExpenseByType(BaseModel):
    expense_type: str
    total: Decimal
    count: int


class ExpenseReport(BaseModel):
    by_type: list[ExpenseByType] = Field(default_factory=list)
    by_date: list[ChartPoint] = Field(default_factory=list)
    total: Decimal

File: frontend/src/types/report.ts
export interface ChartPoint {
  label: string;
  value: string;
}

export interface PaymentStatusBucket {
  status: string;
  count: number;
  revenue: string;
}

export interface SalesReport {
  total_revenue: string;
  total_count: number;
  by_payment_status: PaymentStatusBucket[];
  by_day: ChartPoint[];
}

export interface InventoryReportRow {
  product_id: number;
  name: string;
  sku: string;
  quantity: string;
}

export interface InventoryReport {
  rows: InventoryReportRow[];
  count: number;
}

export interface DebtByCustomer {
  customer_id: number;
  full_name: string;
  remaining: string;
}

export interface DebtByStatus {
  status: string;
  count: number;
  remaining: string;
}

export interface DebtReport {
  by_customer: DebtByCustomer[];
  by_status: DebtByStatus[];
  total_remaining: string;
}

export interface ExpenseByType {
  expense_type: string;
  total: string;
  count: number;
}

export interface ExpenseReport {
  by_type: ExpenseByType[];
  by_date: ChartPoint[];
  total: string;
}

export interface ReportParams {
  store_id?: number;
  date_from?: string;
  date_to?: string;
}
